tests/utils/test_autologging_utils.py

```diff
@@ -0,0 +1,107 @@
+                         ['arg1', 'default1', 'default2'], [1, 2], {}),
+                         (['arg1', 'default1', 'default2'], {},
+                         ['arg1', 'default1', 'default2'], [1, 2], {}),
+                         (['arg1'], {'default1': 42, 'default2': 42},
+                         ['arg1', 'default1', 'default2'], [1, 2], {}),
+                         ([], {'arg1': 42, 'default1': 42, 'default2': 42},
+                         ['arg1', 'default1', 'default2'], [1, 2], {}),
+                         (['user_arg'], {'default1': 42},
+                         ['arg1', 'default1', 'default2'], [1, 2], {'default2': 2}),
+                         (['user_arg'], {'default2': 42},
+                         ['arg1', 'default1', 'default2'], [1, 2], {'default1': 1}),
+                         ([], {'arg1': 42, 'default1': 42},
+                         ['arg1', 'default1', 'default2'], [1, 2], {'default2': 2}),
+                         (['arg1', 'default1'], {},
+                         ['arg1', 'default1', 'default2'], [1, 2], {'default2': 2}),
+                         (['arg1'], {},
+                         ['arg1', 'default1', 'default2'], [1, 2], {'default1': 1, 'default2': 2}),
+                         ([], {'arg1': 42},
+                         ['arg1', 'default1', 'default2'], [1, 2], {'default1': 1, 'default2': 2})]
+
+
+@pytest.mark.large
+@pytest.mark.parametrize("user_args,user_kwargs,all_param_names,all_default_values,expected",
+                         two_default_test_args)
+def test_get_two_unspecified_default_args(user_args, user_kwargs, all_param_names,
+                                          all_default_values, expected):
+
+    default_dict = get_unspecified_default_args(user_args, user_kwargs,
+                                                all_param_names, all_default_values)
+
+    assert default_dict == expected
+
+
+# Test function signature for the following tests
+# def fn_default_default(default1=1, default2=2, default3=3):
+#     pass
+
+
+three_default_test_args = [([], {}, ['default1', 'default2', 'default3'], [1, 2, 3],
+                            {'default1': 1, 'default2': 2, 'default3': 3}),
+                           ([], {'default2': 42},
+                            ['default1', 'default2', 'default3'], [1, 2, 3],
+                            {'default1': 1, 'default3': 3})]
+
+
+@pytest.mark.large
+@pytest.mark.parametrize("user_args,user_kwargs,all_param_names,all_default_values,expected",
+                         three_default_test_args)
+def test_get_three_unspecified_default_args(user_args, user_kwargs, all_param_names,
+                                            all_default_values, expected):
+
+    default_dict = get_unspecified_default_args(user_args, user_kwargs,
+                                                all_param_names, all_default_values)
+
+    assert default_dict == expected
+
+
+@pytest.fixture
+def start_run():
```

Comment:
We should use the `tracking_uri_mock` fixture as an argument to `start_run` to make sure we don't actually log to the local filesystem